File: GCDLib/Core/Visualization/ElevationChangeBarViewer.vb
Imports System.Windows.Forms.DataVisualization

Namespace Core.Visualization

    Public Class ElevationChangeBarViewer

        Public Enum BarTypes
            Area
            Volume
            Vertical
        End Enum

        Private Enum SeriesType
            Erosion
            Depositon
            Net
        End Enum

        Private m_chtControl As Charting.Chart
        'Private _ThresholdedHist As Dictionary(Of Double, Double)
        'Private _RawHist As Dictionary(Of Double, Double)
        Private m_eLinearUnits As UnitsNet.Units.LengthUnit

        Public Sub New(ByRef chtControl As Charting.Chart, eLinearUnits As UnitsNet.Units.LengthUnit)
            m_chtControl = chtControl
            m_eLinearUnits = eLinearUnits

            chtControl.ChartAreas.Clear()
            chtControl.Series.Clear()

            chtControl.ChartAreas.Add("ElevationChangeBars")
            chtControl.Legends.Clear()

            With chtControl.ChartAreas(0).AxisX
                .MajorGrid.Enabled = False
                .MajorTickMark.Enabled = False
            End With

            With chtControl.ChartAreas(0).AxisY
                .MinorTickMark.Enabled = True
                .MajorGrid.LineColor = Drawing.Color.LightSlateGray
                .MinorGrid.Enabled = True
                .MinorGrid.LineColor = Drawing.Color.LightGray
            End With

        End Sub

        Public Sub Refresh(fErosion As Double, fDeposition As Double, eUnits As UnitsNet.Units.LengthUnit, eType As BarTypes, bAbsolute As Boolean)

            Refresh(fErosion, fDeposition, 0, 0, 0, 0, eUnits, False, False, eType, bAbsolute)
        End Sub

        Public Sub Refresh(fErosion As Double, fDeposition As Double, fNet As Double, fErosionError As Double, fDepositionError As Double, fNetError As Double, eUnits As UnitsNet.Units.LengthUnit, eType As BarTypes, bAbsolute As Boolean)

            Refresh(fErosion, fDeposition, fNet, fErosionError, fDepositionError, fNetError, eUnits, True, True, eType, bAbsolute)
        End Sub

        Private Sub Refresh(fErosion As Double, fDeposition As Double, fNet As Double, fErosionError As Double, fDepositionError As Double, fNetError As Double, eUnits As UnitsNet.Units.LengthUnit, bShowErrorBars As Boolean, bShowNet As Boolean, eType As BarTypes, bAbsolute As Boolean)

            m_chtControl.Series.Clear()

            If bAbsolute Then
                ' Bars should have their correct sign. Erosion should be negative
                ' but the number stored in the project is always positive.
                fErosion = -1 * fErosion
            Else
                fNet = Math.Abs(fNet)
            End If

            Dim sYAxisLabel As String = String.Empty
            Select Case eType
                Case BarTypes.Area
                    sYAxisLabel = String.Format("Area ({0}²)", UnitsNet.Length.GetAbbreviation(eUnits))
                Case BarTypes.Volume
                    sYAxisLabel = String.Format("Volume ({0}³)", UnitsNet.Length.GetAbbreviation(eUnits))
                Case BarTypes.Vertical
                    sYAxisLabel = String.Format("Elevation ({0})", UnitsNet.Length.GetAbbreviation(eUnits))
            End Select
            m_chtControl.ChartAreas(0).AxisY.Title = sYAxisLabel

            Dim dSeries As New Dictionary(Of String, Drawing.Color) From {{"Erosion", My.Settings.Erosion}, {"Depsotion", My.Settings.Depsoition}}
            If bShowNet Then
                dSeries.Add("Net", Drawing.Color.Black)
            End If

            Dim errSeries As Charting.Series = m_chtControl.Series.Add("erosion")
            errSeries.Color = Drawing.Color.Red
            errSeries.ChartArea = m_chtControl.ChartAreas.First.Name
            errSeries.ChartType = Charting.SeriesChartType.StackedColumn
            errSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Erosion), fErosion)
            errSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Depositon), 0)

            Dim depSeries As Charting.Series = m_chtControl.Series.Add("deposition")
            depSeries.Color = Drawing.Color.Blue
            depSeries.ChartArea = m_chtControl.ChartAreas.First.Name
            depSeries.ChartType = Charting.SeriesChartType.StackedColumn
            depSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Erosion), 0)
            depSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Depositon), fDeposition)

            If bShowNet Then
                Dim netSeries As Charting.Series = m_chtControl.Series.Add("net")
                netSeries.Color = IIf(fNet >= 0, Drawing.Color.Blue, Drawing.Color.Red)
                netSeries.ChartArea = m_chtControl.ChartAreas.First.Name
                netSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Erosion), 0)
                netSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Depositon), 0)
                netSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Net), fNet)

                errSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Net), 0)
                depSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Net), 0)
            End If

            m_chtControl.ChartAreas(0).RecalculateAxesScale()
            m_chtControl.AlignDataPointsByAxisLabel()

        End Sub

        Private Function GetXAxisLabel(eBarType As BarTypes, eSeriesType As SeriesType)

            Dim sBarType As String = String.Empty
            Select Case eBarType
                Case BarTypes.Area
                    sBarType = "Total\nArea"
                Case BarTypes.Volume
                    sBarType = "Total\nVolume"
                Case BarTypes.Vertical
                    sBarType = "Average\nDepth"
            End Select

            Dim sSeriesType As String = String.Empty
            Select Case eSeriesType
                Case SeriesType.Erosion
                    sSeriesType = "Erosion"
                Case SeriesType.Depositon
                    sSeriesType = "Deposition"
                Case SeriesType.Net
                    If eBarType = BarTypes.Volume Then
                        Return String.Format("Total{0}Net Volume{0}Difference", Environment.NewLine)
                    ElseIf eBarType = BarTypes.Vertical Then
                        Return String.Format("Avg. Total{0}Thickness{0}Difference", Environment.NewLine)
                    End If
            End Select

            Return String.Format("{1} of{0}{2}", Environment.NewLine, sBarType, sSeriesType)

        End Function

        Public Sub Save(sFilePath As String, nChartWidth As Integer, nChartHeight As Integer, nDPI As Integer)
            m_chtControl.SaveImage(sFilePath, Charting.ChartImageFormat.Png)
        End Sub

    End Class

End Namespace

File: GCDCore/Visualization/ElevationChangeBarViewer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Drawing;
using System.Windows.Forms.DataVisualization.Charting;

namespace GCDCore.Visualization
{
    public class ElevationChangeBarViewer : ViewerBase
    {
        public enum BarTypes
        {
            Area,
            Volume,
            Vertical
        }

        private enum SeriesType
        {
            Erosion,
            Depositon,
            Net
        }

        public ElevationChangeBarViewer(Chart chtControl = null)
            : base(chtControl)
        {
            var _with1 = Chart.ChartAreas[0].AxisX;
            _with1.MajorGrid.Enabled = false;
            _with1.MajorTickMark.Enabled = false;

            var _with2 = Chart.ChartAreas[0].AxisY;
            _with2.MinorTickMark.Enabled = true;
            _with2.MajorGrid.LineColor = Color.LightSlateGray;
            _with2.MinorGrid.Enabled = true;
            _with2.MinorGrid.LineColor = Color.LightGray;

            Chart.Series.Clear();

            Series errSeries = Chart.Series.Add(ViewerBase.EROSION);
            errSeries.Color = Properties.Settings.Default.Erosion;
            errSeries.ChartArea = Chart.ChartAreas.First().Name;
            errSeries.ChartType = SeriesChartType.StackedColumn;

            Series depSeries = Chart.Series.Add(ViewerBase.DEPOSITION);
            depSeries.Color = Properties.Settings.Default.Deposition;
            depSeries.ChartArea = Chart.ChartAreas.First().Name;
            depSeries.ChartType = SeriesChartType.StackedColumn;
        }

        public void Refresh(double fErosion, double fDeposition, string sDisplayUnitsAbbreviation, BarTypes eType, bool bAbsolute)
        {
            Refresh(fErosion, fDeposition, 0, 0, 0, 0, sDisplayUnitsAbbreviation, false, false, eType, bAbsolute);
        }

        public void Refresh(double fErosion, double fDeposition, double fNet, double fErosionError, double fDepositionError, double fNetError, string sDisplayUnitsAbbreviation, BarTypes eType, bool bAbsolute)
        {
            Refresh(fErosion, fDeposition, fNet, fErosionError, fDepositionError, fNetError, sDisplayUnitsAbbreviation, true, true, eType, bAbsolute);
        }

        private void Refresh(double fErosionOrig, double fDeposition, double fNetOrig, double fErosionError, double fDepositionError, double fNetError,
            string sDisplayUnitsAbbreviation, bool bShowErrorBars, bool bShowNet, BarTypes eType, bool bAbsolute)
        {
            double max, min;

            // We don't want to mutate our parameters so
            double fNet = fNetOrig;
            double fErosion = fErosionOrig;

            if (bAbsolute)
            {
                // Bars should have their correct sign. Erosion should be negative
                // but the number stored in the project is always positive.
                fErosion = -1 * fErosion;
            }
            else
            {
                fNet = Math.Abs(fNet);
            }

            string sYAxisLabel = string.Empty;
            switch (eType)
            {
                case BarTypes.Area:
                    sYAxisLabel = string.Format("Area ({0})", sDisplayUnitsAbbreviation);
                    break;
                case BarTypes.Volume:
                    sYAxisLabel = string.Format("Volume ({0})", sDisplayUnitsAbbreviation);
                    break;
                case BarTypes.Vertical:
                    sYAxisLabel = string.Format("Elevation ({0})", sDisplayUnitsAbbreviation);
                    break;
            }
            Chart.ChartAreas[0].AxisY.Title = sYAxisLabel;

            Dictionary<string, Color> dSeries = new Dictionary<string, Color> {
                {
                    "Lowering",
                    Properties.Settings.Default.Erosion
                },
                {
                    "Raising",
                    Properties.Settings.Default.Deposition
                }
            };

            if (bShowNet)
            {
                dSeries.Add("Net", Color.Black);
            }

            Series errSeries = Chart.Series[ViewerBase.EROSION];
            errSeries.Points.Clear();
            errSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Erosion), fErosion);
            errSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Depositon), 0);

            Series depSeries = Chart.Series[ViewerBase.DEPOSITION];
            depSeries.Points.Clear();
            depSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Erosion), 0);
            depSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Depositon), fDeposition);

            max = Math.Max(fErosion, fDeposition);
            min = Math.Min(Math.Min(fErosion, fDeposition), 0);

            Series netSeries = Chart.Series.FindByName(ViewerBase.NET);
            if (bShowNet)
            {
                if (netSeries == null)
                    netSeries = Chart.Series.Add(ViewerBase.NET);

                netSeries.Points.Clear();

                netSeries.Color = (fNetOrig >= 0 ? depSeries.Color : errSeries.Color);
                netSeries.ChartArea = Chart.ChartAreas.First().Name;
                netSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Erosion), 0);
                netSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Depositon), 0);
                netSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Net), fNet);

                errSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Net), 0);
                depSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Net), 0);

                max = Math.Max(max, fNet);
                min = Math.Min(min, fNet);
            }
            else
            {
                if (netSeries is Series)
                    Chart.Series.Remove(netSeries);
            }

            Series errorSeries = Chart.Series.FindByName(ViewerBase.ERROR);
            if (eType != BarTypes.Area)
            {
                if (errorSeries == null)
                {
                    errorSeries = Chart.Series.Add(ViewerBase.ERROR);
                    errorSeries.ChartType = SeriesChartType.ErrorBar;
                    errorSeries.Color = Color.Black;
                }

                errorSeries.Points.Clear();

                errorSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Erosion), fErosion, fErosion - fErosionError, fErosion + fErosionError);
                errorSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Depositon), fDeposition, fDeposition - fDepositionError, fDeposition + fDepositionError);

                max = Math.Max(fErosion + fErosionError, fDeposition + fDepositionError);
                min = Math.Min(0, Math.Min(fErosion - fErosionError, fDeposition - fDepositionError));

                if (netSeries is Series)
                    errorSeries.Points.AddXY(GetXAxisLabel(eType, SeriesType.Net), fNet, fNet - fNetError, fNet + fNetError);

                max = Math.Max(max, fNet + fNetError);
                min = Math.Min(min, fNet - fNetError);
            }
            else
            {
                if (errorSeries is Series)
                    Chart.Series.Remove(errorSeries);
            }

            try
            {
                // Set an initial max and min so the autoscale can do its thing
                Chart.ChartAreas[0].AxisY.Maximum = max;
                Chart.ChartAreas[0].AxisY.Minimum = min;
                Chart.ChartAreas[0].RecalculateAxesScale();

                Tuple<decimal, decimal> maxmin = GCDConsoleLib.Utility.IntervalMath.GetRegularizedMaxMin((decimal)max, (decimal)min, 0.02m);

                Chart.ChartAreas[0].AxisY.Maximum = (double)maxmin.Item1;
                Chart.ChartAreas[0].AxisY.Minimum = (double)maxmin.Item2;

                Chart.ChartAreas[0].AxisY.Interval = (double)GCDConsoleLib.Utility.IntervalMath.GetSensibleChartInterval(
                    (decimal)Chart.ChartAreas[0].AxisY.Maximum,
                    (decimal)Chart.ChartAreas[0].AxisY.Minimum,
                    10);

                Chart.AlignDataPointsByAxisLabel();
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error calculating change bar chart axes: " + ex.Message);
                //throw new Exception("Error refreshing elevation bar charts.", ex);
            }
        }


        private object GetXAxisLabel(BarTypes eBarType, SeriesType eSeriesType)
        {
            string sBarType = string.Empty;
            switch (eBarType)
            {
                case BarTypes.Area:
                    sBarType = "Total\\nArea";
                    break;
                case BarTypes.Volume:
                    sBarType = "Total\\nVolume";
                    break;
                case BarTypes.Vertical:
                    sBarType = "Average\\nDepth";
                    break;
            }

            string sSeriesType = string.Empty;
            switch (eSeriesType)
            {
                case SeriesType.Erosion:
                    sSeriesType = "Lowering";
                    break;
                case SeriesType.Depositon:
                    sSeriesType = "Raising";
                    break;
                case SeriesType.Net:
                    if (eBarType == BarTypes.Volume)
                    {
                        return string.Format("Total{0}Net Volume{0}Difference", Environment.NewLine);
                    }
                    else if (eBarType == BarTypes.Vertical)
                    {
                        return string.Format("Avg. Total{0}Thickness{0}Difference", Environment.NewLine);
                    }
                    break;
            }

            return string.Format("{1} of{0}{2}", Environment.NewLine, sBarType, sSeriesType);

        }

        public void Save(System.IO.FileInfo filePath, int nChartWidth, int nChartHeight)
        {
            try
            {
                SaveImage(filePath);
            }
            catch(Exception ex)
            {
                Console.WriteLine("Error saving elevation change bar chart: " + ex.Message);
            }
        }
    }
}